rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll

```diff
@@ -117,51 +117,63 @@ class BooleanCompletion extends ConditionalCompletion, TBooleanCompletion {
   override string toString() { result = "boolean(" + value + ")" }
 }
 
-/** Holds if `pat` is guaranteed to match at the point in the AST where it occurs. */
-pragma[nomagic]
-private predicate isExhaustiveMatch(Pat pat) {
-  (
-    pat instanceof WildcardPat
-    or
-    pat = any(IdentPat ip | not ip.hasPat() and ip = any(Variable v).getPat())
-    or
-    pat instanceof RestPat
-    or
-    // `let` statements without an `else` branch must be exhaustive
-    pat = any(LetStmt let | not let.hasLetElse()).getPat()
-    or
-    // `match` expressions must be exhaustive, so last arm cannot fail
-    pat = any(MatchExpr me).getLastArm().getPat()
-    or
-    // macro invocations are exhaustive if their expansion is
-    pat = any(MacroPat mp | isExhaustiveMatch(mp.getMacroCall().getExpanded()))
-    or
-    // parameter patterns must be exhaustive
-    pat = any(Param p).getPat()
-  ) and
-  not pat = any(ForExpr for).getPat() // workaround until `for` loops are desugared
+/**
+ * Holds if `pat` can not _itself_ be the cause of a pattern match failure. This
+ * does not mean that `pat` is irrefutable, as its children might be the cause
+ * of a failure.
```

Comment:
This QL doc seems to suggest the dual of what the predicate actually computes? I wonder if it is actually better to implement what the QL doc says, e.g.
```
private predicate cannotCauseMatchFailure(Pat pat) {
  pat = any(IdentPat p | p.hasPat())
  or
  pat instanceof BoxPat
  or
  pat instanceof RestPat
  or
  pat instanceof MacroPat
  or
  pat instanceof WildcardPat
  or
  pat instanceof RangePat
  or
  pat instanceof RefPat
  or
  pat instanceof SlicePat
  or
  pat instanceof TuplePat
  or
  pat instanceof ConstBlockPat
}
```

and then replace `not canCauseMatchFailure(pat)` with `cannotCauseMatchFailure(pat)` further down.